feat: Conversation context menu

fayazara opened the pull request:
## Description
Demo
https://www.loom.com/share/b18f0792fb044807a3fb01fa03491094

A right click menu on conversation to quickly resolve, reopen, mark as pending, snooze convresations and also assign agents and labels.

Fixes #5074

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

nithindavid commented on app/javascript/dashboard/components/widgets/conversation/contextMenu/Index.vue:
```diff
@@ -0,0 +1,264 @@
+  min-width: calc(var(--space-mega) * 2);
+  position: relative;
+  border-radius: var(--border-radius-medium);
+  padding: var(--space-half);
```
Can we stick to 4,8,16 for spacing ?

nithindavid commented on app/javascript/dashboard/components/widgets/conversation/contextMenu/Index.vue:
```diff
@@ -0,0 +1,264 @@
+}
+
+.menu-item {
+  padding: var(--space-half);
```
--space-small or smaller

nithindavid commented on app/javascript/dashboard/components/widgets/conversation/contextMenu/Index.vue:
```diff
@@ -0,0 +1,264 @@
+      border: 1px solid var(--s-25);
+      min-width: calc(var(--space-mega) * 2);
+      border-radius: var(--border-radius-medium);
+      padding: var(--space-half);
```
Spacing small or smaller

nithindavid reviewed (changes requested):
Have left some minor design changes. Please see them

fayazara commented:
> Have left some minor design changes. Please see them

Updated @nithindavid

fayazara commented on app/javascript/dashboard/components/widgets/conversation/ConversationCard.vue:
```diff
@@ -264,6 +286,14 @@ export default {
+    },
+  },
+  mounted() {
+    this.$store.dispatch('inboxAssignableAgents/fetch', [this.inbox.id]);
```
> This is not the correct place for this call. It would create unnecessary calls. The fetch call should be fired only if the context menu is opened.
> 
> <img alt="Screenshot 2022-07-20 at 3 00 29 PM" width="1125" src="https://user-images.githubusercontent.com/2246121/179949151-fb966306-b57e-4371-a139-ba5db124fac5.png">

I thought about this, but the way the submenu is show is purely via css, hence I needed a way to hook the api call on a mounted method.

Will figure out a way to handle this some other way.

pranavrajs commented on app/javascript/dashboard/components/widgets/conversation/contextMenu/Index.vue:
```diff
@@ -0,0 +1,264 @@
+  <div class="conversation-menu">
+    <!-- // Status Menu  -->
+    <template v-for="item in menuItems">
+      <div
+        v-if="show(item.key)"
+        :key="item.key"
+        class="menu-item flex-align-center"
+        @click="toggleStatus(item.key, null)"
+      >
+        <fluent-icon :icon="item.icon" size="14" class="icon" />
+        <p>{{ item.label }}</p>
+      </div>
```
Make this a component. (context-menu-item)

pranavrajs commented on app/javascript/dashboard/components/widgets/conversation/contextMenu/Index.vue:
```diff
@@ -0,0 +1,264 @@
+    </template>
+    <!-- // Status Menu  -->
+    <!-- // Snooze Menu  -->
+    <div class="menu-item has-submenu">
+      <div class="flex-align-center">
+        <fluent-icon icon="snooze" size="14" class="icon" />
+        <p>{{ $t('CONVERSATION.CARD_CONTEXT_MENU.SNOOZE.TITLE') }}</p>
+      </div>
+      <fluent-icon icon="chevron-right" size="12" />
+      <div class="conversation-submenu">
+        <div class="menu-item" @click="toggleStatus(STATUS_TYPE.SNOOZED, null)">
```
Make this a a new component (context-menu-with-submenu). Use the menu-item created previously for submenu.

pranavrajs commented on app/javascript/dashboard/components/widgets/conversation/contextMenu/Index.vue:
```diff
@@ -0,0 +1,264 @@
+    </div>
+    <!-- // Agent Menu  -->
+    <!-- // Label Menu  -->
+    <div class="menu-item has-submenu">
```
Use the component created above.

pranavrajs commented on app/javascript/dashboard/components/widgets/conversation/contextMenu/Index.vue:
```diff
@@ -0,0 +1,264 @@
+    ...mapGetters({
+      labels: 'labels/getLabels',
+    }),
+    snoozeTimes() {
+      return {
+        // tomorrow  = 9AM next day
+        tomorrow: getUnixTime(addHours(startOfTomorrow(), 9)),
+        // next week = 9AM Monday, next week
+        nextWeek: getUnixTime(
+          addHours(startOfWeek(addWeeks(new Date(), 1), { weekStartsOn: 1 }), 9)
+        ),
+      };
+    },
```
It feels like this is repeated. Please create a mixin.

pranavrajs reviewed (changes requested):
Nice work! Looks good overall. I have some comments on the components and the actions used. Please take a look at the individual comments.

iamsivin commented:
@fayazara What about adding icons here also?

![image](https://user-images.githubusercontent.com/64252451/180694277-ce6d0f2c-1cf0-4561-addf-2a5c344e465b.png)

fayazara commented:
> @fayazara What about adding icons here also?

@iamsivin This was mostly kept text only, because I couldn't find icons which were linked to those keywords. Hence keeping just plain text made sense.

fayazara commented:
> Nice work! Looks good overall. I have some comments on the components and the actions used. Please take a look at the individual comments.

@pranavrajs 
Refactored the entire module.

1. Removed all the actions from `ConversationCard` component to keep it dumb.
2. All events are not being relayed to `Chatlist.vue` and I am using the existing methods for assign agents/labels and updating status, except Snoozed - because I am not sure if our bulk actions apis supports snooze timestamp yet.
3. Fixed the Firefox scrollbar bug - adding `overflow: hidden` to conversations list, this will also help reduce confusion when a menu is opened and is a user scrolls down, the menu was staying there with a fixed position.
4. Added mixins for snoozeTimes.
